pxr/imaging/hdSt/materialXFilter.cpp

```diff
@@ -135,6 +190,135 @@ _CreateHdStMaterialXContext(
+
+    size_t nodeCounter = 0;
+    nodePathMap->clear();
+
+    // Paths will go /NG_Anonymized/N0, /NG_Anonymized/N1, /NG_Anonymized/N2...
+    SdfPath ngBase(_tokens->NG_Anonymized);
+
+    // We will traverse the network in a depth-first traversal starting at the
+    // terminals. This will allow a stable traversal that will not be affected
+    // by the ordering of the SdfPaths and make sure we assign the same index to
+    // all nodes regardless of the way they are sorted in the network node map.
+    std::vector<const SdfPath*> pathsToTraverse;
+    for (const auto& terminal : hdNetwork.terminals) {
+        const auto& connection = terminal.second;
+        pathsToTraverse.push_back(&(connection.upstreamNode));
+    }
+    while (!pathsToTraverse.empty()) {
+        const SdfPath* path = pathsToTraverse.back();
+        pathsToTraverse.pop_back();
+        if (!(*nodePathMap).count(*path)) {
+            const HdMaterialNode2& node = hdNetwork.nodes.find(*path)->second;
+            // We only need to create the anonymized name at this time:
+            (*nodePathMap)[*path] = ngBase.AppendChild(TfToken("N" + std::to_string(nodeCounter++)));
+            for (const auto& input : node.inputConnections) {
+                for (const auto& connection : input.second) {
+                    pathsToTraverse.push_back(&(connection.upstreamNode));
+                }
+            }
+        }
+    }
+
+    // Copy the incoming network using only the anonymized names:
+    topoNetwork->primvars = hdNetwork.primvars;
+    for (const auto& terminal : hdNetwork.terminals) {
+        topoNetwork->terminals.emplace(
+            terminal.first,
+            HdMaterialConnection2 { (*nodePathMap)[terminal.second.upstreamNode],
+                                    terminal.second.upstreamOutputName });
+    }
+    for (const auto& nodePair : hdNetwork.nodes) {
+        const HdMaterialNode2& inNode = nodePair.second;
+        HdMaterialNode2        outNode;
+        outNode.nodeTypeId = inNode.nodeTypeId;
+        if (_IsTopologicalShader(inNode.nodeTypeId)) {
+            // Topological nodes have parameters that affect topology. We can not strip
+            // them.
+            outNode.parameters = inNode.parameters;
+        } else {
+            // Parameters that are color managed are also topological as they
+            // result in different nodes being added in the MaterialX graph
+            for (const auto& param: inNode.parameters) {
+                const auto colorManagedFile = 
+                    SdfPath::StripPrefixNamespace(param.first.GetString(),
+                                                  SdfFieldKeys->ColorSpace);
+                if (colorManagedFile.second) {
+                    outNode.parameters.insert(param);
+                    // Need an empty asset as well:
+                    outNode.parameters.emplace(TfToken(colorManagedFile.first),
+                                               VtValue{SdfAssetPath{}});
```

Comment:
Might be incorrect if there are plans to color-correct color3 values on nodes that are not file based.

Reply:
We're planning to, in some form, but I think we can update this bit when we get to it.

Reply:
We actually already support that for MaterialX - there is an existing test case but it looks like the opensource thresholds were not tight enough to catch this. Let me know if you are good with this change to catch color corrected  values
```
const auto colorManagedInput = 
SdfPath::StripPrefixNamespace(param.first.GetString(), SdfFieldKeys->ColorSpace);

// If this parameter indicates a colorspace on a color 
// managed input, find and add that corresponding input
if (colorManagedInput.second) {
    outNode.parameters.insert(param);

    // Get the parameter value for the input
    const TfToken colorInputParam(colorManagedInput.first);
    const auto colorInputIt = inNode.parameters.find(colorInputParam);
    if (colorInputIt != inNode.parameters.end()) {
        VtValue colorInputValue = colorInputIt->second;

        // Use an empty asset for color managed files
        if (colorInputValue.IsHolding<SdfAssetPath>()) {
            colorInputValue = VtValue(SdfAssetPath());
        }
        outNode.parameters.emplace(colorInputParam, colorInputValue); 
    }
}
```

Reply:
Sadly there will not be a value to be found in the data since it all gets pruned to create a pure "default" shader. But the answer we are looking for can be found in the Sdr registry with code pretty close to this:

```C++
const auto colorManagedInput = 
SdfPath::StripPrefixNamespace(param.first.GetString(), SdfFieldKeys->ColorSpace);

// If this parameter indicates a colorspace on a color 
// managed input, we need to find out if it manages an asset.
if (colorManagedInput.second) {
    outNode.parameters.insert(param);

    // Get the parameter type for the input
    const TfToken colorInputParam(colorManagedInput.first);
    SdrRegistry &sdrRegistry = SdrRegistry::GetInstance();
    const SdrShaderNodeConstPtr sdrNode = 
        sdrRegistry.GetShaderNodeByIdentifierAndType(inNode.nodeTypeId, _tokens->mtlx);
    auto sdrInput = sdrNode->GetShaderInput(colorInputParam);
    if (sdrInput->GetTypeAsSdfType().first == SdfValueTypeNames->Asset) {
        // Use an empty asset for color managed files
        colorInputValue = VtValue(SdfAssetPath());
        outNode.parameters.emplace(colorInputParam, colorInputValue); 
    }
}
```

The other solution would be to fix `_AddMaterialXNode()` in `usd\pxr\imaging\hdMtlx\hdMtlx.cpp` to change this block:

```C++
        // Skip Colorspace parameter, this is already captured in the paramData.
        // Note: Colorspace inputNames are of the form 'colorSpace:inputName'
        const std::pair<std::string, bool> result = 
            SdfPath::StripPrefixNamespace(mxInputName, SdfFieldKeys->ColorSpace);
        if (result.second) {
            continue;
        }
```

to:

```C++
        // Colorspace parameter: create an empty input if it does not already
        // exist to set colorspace metadata.
        // Note: Colorspace inputNames are of the form 'colorSpace:inputName'
        const std::pair<std::string, bool> result = 
            SdfPath::StripPrefixNamespace(mxInputName, SdfFieldKeys->ColorSpace);
        if (result.second) {
            mx::InputPtr mxInput = mxNode->addInputFromNodeDef(result .first);
            mxInput->setColorSpace(HdMtlxConvertToString(paramData.value));
            continue;
        }
```

Which would allow not having to create a dummy value in the first place.

Reply:
Actually had to fix this on our side since it impacted our internal unit tests. The following function was added:
```C++
// MaterialX requires a value for color-spaced attributes. Fetch the default from Sdr shader:
void _AddMaterialXDefaultValue(HdMaterialNode2& node, std::string const& parameterName) {
    SdrRegistry &sdrRegistry = SdrRegistry::GetInstance();
    const SdrShaderNodeConstPtr sdrNode = sdrRegistry.GetShaderNodeByIdentifierAndType(
        node.nodeTypeId, _tokens->mtlx);
    if (sdrNode) {
        SdrShaderPropertyConstPtr sdrInput = sdrNode->GetShaderInput(TfToken(parameterName));
        if (sdrInput) {
            if (sdrInput->IsAssetIdentifier()) {
                // Replace empty filename with a placeholder ignored by MaterialX shadergen:
                node.parameters.emplace(TfToken(parameterName),
                                        VtValue{SdfAssetPath{"placeholder"}});
            } else {
                node.parameters.emplace(TfToken(parameterName),
                                        sdrInput->GetDefaultValueAsSdfType());
            }
        }
    }
}
```
and was called:
```diff
@@ -282,9 +303,8 @@ size_t _BuildEquivalentMaterialNetwork(
                                                   SdfFieldKeys->ColorSpace);
                 if (colorManagedFile.second) {
                     outNode.parameters.insert(param);
-                    // Need an empty asset as well:
-                    outNode.parameters.emplace(TfToken(colorManagedFile.first),
-                                               VtValue{SdfAssetPath{}});
+                    // We need a value as well:
+                    _AddMaterialXDefaultValue(outNode, colorManagedFile.first);
                 }
             }
         }
```
And this made sure there was a correct value for the MaterialX code.